jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java

```diff
@@ -0,0 +1,735 @@
+        }
+    }
+
+    /**
+     * A getter to obtain element values of Vector data structures.
+     * 
+     * @author Hannes Wellmann
+     *
+     * @param <V> the type of handled vectors
+     */
+    public interface ElementGetter<V>
+    {
+        /**
+         * Returns the value of the element at the specified {@code index} in the given {@link V
+         * vector}.
+         * 
+         * @param vector from which an element value is to return
+         * @param index of the element to return
+         * @return the value of the specified element in the given vector
+         */
+        double getElement(V vector, int index);
+    }
+
+    private final Function<double[], V> vectorFactory;
+    private final ElementGetter<V> elementGetter;
+    private int vectorLength = -1;
+
+    /**
+     * Create a {@link TSPLIBImporter} that uses the given {@code vectorFactory} to create its
+     * vertices with certain element values and later uses the specified {@code elementGetter} to
+     * obtain the element values and compute the distances between the vertices.
+     * 
+     * @param vectorFactory used to create vector vertices with certain element values
+     * @param elementGetter used to obtain element values from the created vertex-objects
+     */
+    public TSPLIBImporter(Function<double[], V> vectorFactory, ElementGetter<V> elementGetter)
+    {
+        this.vectorFactory = vectorFactory;
+        this.elementGetter = elementGetter;
+    }
+
+    // import data
+
+    private TSPLIBFileData<V> lastImportData = null;
+
+    /**
+     * Returns the {@link TSPLIBFileData} of the latest imported file or null, if no import
+     * completed yet or the latest import failed.
+     * 
+     * @return {@code TSPLIBFileData} of the latest import
+     */
+    public TSPLIBFileData<V> getLastImportData()
+    {
+        return lastImportData;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * The given {@link Graph}, if not null, must have a {@link GraphType#isSimple() simple type}
+     * and must be weighted and empty. If the given {@code Graph} is null, a suitable one is
+     * created.
+     * </p>
+     * <p>
+     * This implementation is not thread-safe and must be synchronized externally if called by
+     * concurrent threads.
+     * </p>
+     * <p>
+     * {@link TSPLIBFileData} of the previous import can be obtained from
+     * {@link #getLastImportData()}.
+     * </p>
+     * 
+     * @param graph into which this importer writes, must be simple, weighted and empty or may be
+     *        null.
+     * @throws IllegalArgumentException if the specified {@code graph} is not simple or is not empty
+     */
+    @Override
+    public void importGraph(Graph<V, DefaultWeightedEdge> graph, Reader in)
+    {
+        lastImportData = null;
+        vectorLength = -1;
+        if (graph == null) {
+            @SuppressWarnings("unchecked") Supplier<DefaultWeightedEdge> edgeSupplier =
```

Comment:
The graph is not supposed to be null here. It is not your responsibility to create a graph here.

Now I understand why you return the graph inside the metadata object. Please remove this pattern. We want to keep all importers working in a similar way.